client/src/redux/hooks.js

```diff
@@ -1,54 +1,123 @@
-import { useCallback } from 'react';
-import { useDispatch, shallowEqual, useSelector } from 'react-redux';
-import { loadUser, register, login, logout } from './actions/auth';
+import { useCallback } from "react";
+import { useDispatch, shallowEqual, useSelector } from "react-redux";
+import { loadUser, register, login, logout } from "./actions/auth";
+import {
+  getProverbs,
+  addProverb,
+  deleteProverb,
+  updateProverb,
+  getProverb,
+} from "./actions/proverb";
 
 export function useAuth() {
-    const dispatch = useDispatch();
-    const { token, isAuthenticated, loading, user } = useSelector(
-      (state) => ({
-        token: state.auth.token,
-        isAuthenticated: state.auth.isAuthenticated,
-        loading: state.auth.loading,
-        user: state.auth.user
-      }),
-      shallowEqual
-    );
-  
-    const boundLoadUser = useCallback(
-      (...args) => {
-        return dispatch(loadUser(...args));
-      },
-      [dispatch]
-    );
-  
-    const boundRegisterUser = useCallback(
-      (...args) => {
-        return dispatch(register(...args));
-      },
-      [dispatch]
-    );
-  
-    const boundLoginUser = useCallback(
-      (...args) => {
-        return dispatch(login(...args));
-      },
-      [dispatch]
-    );
-    const boundLogoutUser = useCallback(
-      (...args) => {
-        return dispatch(logout(...args));
-      },
-      [dispatch]
-    );
-  
-    return {
-      token,
-      isAuthenticated,
-      loading,
-      user,
-      loadUser: boundLoadUser,
-      registerUser: boundRegisterUser,
-      loginUser: boundLoginUser,
-      logoutUser: boundLogoutUser
-    };
-  }
\ No newline at end of file
+  const dispatch = useDispatch();
+  const { token, isAuthenticated, loading, user } = useSelector(
```

Comment:
If you create selector functions in a different file, they become easier to reuse. for example, you could have a file called "authSelectors" an there, all the selectors related to getting information from the auth state property. And here you could do:

```
const {token, isAuthenticated, loading, user} = useSelector(selectAuth) 
```

Also, I would say to avoid retuning new objects from selectors although sometimes could be necessary. The reason is, when you return new arrays or objects, you also are returning a new reference of them each time the selector is invoked (that's why you had to use the shallowEqual). It's very easy to forget the shallowEqual, then you would have extra renders all the time.

Reply:
If you want to extract multiple values from the state you have to return an object, otherwise We have to call `useSelector` multiple times, as you mentioned that is why we should use `shallowEqual` to avoid rerender. 

According to the official docs on [Redux page](https://react-redux.js.org/api/hooks)

> You may call useSelector() multiple times within a single function component. Each call to useSelector() creates an individual subscription to the Redux store. Because of the React update batching behavior used in React Redux v7, a dispatched action that causes multiple useSelector()s in the same component to return new values should only result in a single re-render.

I think, we can call `useSelector` multiple times in the component so that it will only cause one render because it uses strict equality and returns a primitive value. However, I would say if we dont have a deep nested and large object, I will prefer using `shallowEqual` :)

Reply:
You are right and I am glad you research it: sometimes it is necessary to return new objects or arrays from the selectors. Then you have to use the shallowEqual.

However, it is good to keep those cases when it is necessary and there is no other option. 

In this specific case, you are returning all the properties from the auth state so selectAuth selector could be simplified to this:

Current implementation:
```
export const selectAuth =  (state) => ({
    token: state.auth.token,
    isAuthenticated: state.auth.isAuthenticated,
    loading: state.auth.loading,
    user: state.auth.user,
  })
```

Suggestion change: 

```
export const selectAuth =  (state) => state.auth
```

With this implementation, you wouldn't need to use the shallow equal.

Reply:
Frank, Thank you for your answer and your time. However I want to share some links, I think some developers say that it cause also rerender. Could you look at it?

`  const { token, isAuthenticated, loading, user } = useSelector(
    selectAuth)
`


 [ github](https://github.com/reduxjs/react-redux/issues/1459) 

 [ medium](https://medium.com/swlh/how-to-make-useselector-not-a-disaster-6fdd1c558fd3)

Reply:
I create a small example to show a bit deeper the behaviour of the useSelector and when it forces the component to re-render.

https://codesandbox.io/s/funny-mayer-ztzqt?file=/src/index.js

We can talk about it on Saturday if you want

Cheers.

Reply:
very nice that you are looking for information!

if you check both articles, they are talking about return new object from the selectors. There is a big difference between this two selectors: 

```
function selectorA(state) {
    return {
         token: state.auth.token,
         loading: state.auth.loading
    }
}
```

```
function selectorB(state) {
    return state.auth
}
```

Let's talk about it on Saturday!

Cheers.